capi-adoption/README.md

```diff
@@ -0,0 +1,114 @@
+
+### Versioning of CRDs
+
+Cluster API is still in a very early stage of development and its CRD versions change once or twice a year. Kubernetes api-server can only manage one CRD version at a time as a storage version and backwards and upwards support for alternative versions must be implemented by appropriate webhooks. Depending on the change in a CRD structure these webhooks work or not.
+
+Currently the stable API version for Cluster API is v1alpha3 but v1alpha4 is already well under development. It is highly possible that when we get our first production ready version of GS managed Cluster API released, we soon must migrate to the next one involving CRD migration. This can become very problematic when the management cluster has a large number of active releases that contain components that utilize different versions of CRDs.
+
+### Operator reconciliation approach towards resource management
+
+At Giant Swarm a very unique thing we do with these cluster related controllers is that deploying a version of the controllers in most cases leads to a direct impact on the infrastructure resources it is managing in the form of an upgrade.
+This is very different from upstream and other community operators where usually an upgrade of the controller SHOULD NOT touch anything on the resource side, but any resource change is usually explicitly affected by a change in a field or annotation, e.g. by changing the prometheus version in a CR that prometheus operator manages or by changing the kubernetes version in a controlplane CR for CAPI.
+
+This is also more in line with native K8s controllers, where a new Deployment controller does not roll all deployments for example. It might offer new opt-in features that you can use, but you have to do that explicitly.
+
+### Immutable release of all components
+
+As briefly mentioned earlier in this document, the Giant Swarm releases are currently strictly defining all specific components that are either needed to create a workload cluster or that are running on the workload cluster. This includes operators running on management cluster, used Flatcar image version, installed Kubernetes version and several apps for both management cluster and workload cluster.
+
+When we move to fully use upstream Cluster API controllers to manage workload clusters, the whole concept of versioning changes radically. With the Cluster API CRDs and the upstream controllers, the end-user gets a much more flexible way to create, configure and manage workload clusters. At the same time the mechanisms how different components are configured, changes dramatically.
```

Comment:
I think this is the crux of the whole discussion:
We talk about how Cluster API manages the clusters and infrastructure but at least to me it's not clear what the "upstream" way is we describe here.
There are several different ways to run CAPI: From what I gathered this document aims at the "replace all controllers" in place and always keep them on latest style.

Reply:
1. Yes, we replace all controllers in place and keep them in latest stable releases. In other words: We always update to latest patches on given MC.
2. The versioning changes radically because workload cluster kubernetes version, etcd version or given in-cluster DNS application / its version are not coupled with the version of operators. This is extremely important improvement because now customers can test different kubernetes versions without getting changes to infrastructure at the same time.

Reply:
1. I get that but it's not clearly stated in the RFC doc so far. See also @JosephSalisbury confusion further down as I got the impression that he was talking about the whole "one operator version per MC and then move CRs between MCs approach".

2. I would be questioning the limitations here. Upstream was recently discussing the range of k8s versions they support (generally the last 3 IIRC). How do we convey this to the customer without having a relation between these versions (e.g. a release)? Similarly the problem exists with the new latest kubernetes versions. CAPI AFAIK is not immediately compatible with a new version but did need _some_ (not much) time to bring out a new patch to support it.
We should think about how we convey this and structure this. I agree that it needs to be build into defaulting / validation but it will still be a problem.

Reply:
Regarding limits or supported range of versions for point 2., I consider it rather as a matter of communication than a technical issue. When we don't strictly control the support but just provide what upstream provides, I see it similar to kubernetes features in general: Customers need to educate themselves in order to know what is possible with given versions.

Reply:
Further with the latest improvements in App Platform, we could restrict the version of apps based on k8s version. Also with `AppSets` we could wrap a collection of apps certified/tested by us. I think this step (point 2) is a huge step forward the right direction